src/auth/mod.rs

```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+mod context;
+mod session;
+mod sessions;
+mod sid;
+mod user;
+
+pub use self::user::User;
+pub use openidconnect::url::Url;
+
+use std::fmt::Display;
+use std::time::Duration;
+
+use axum::extract::{Extension, Query};
+use axum::headers::HeaderName;
+use axum::http::{HeaderValue, StatusCode};
+use axum::response::{IntoResponse, Redirect};
+use axum::routing::get;
+use axum::Router;
+
+use openidconnect::core::{CoreClient, CoreProviderMetadata};
+use openidconnect::reqwest::async_http_client;
+use openidconnect::{AuthType, ClientId, ClientSecret, IssuerUrl, RedirectUrl};
+
+use anyhow::{Context as _, Error};
+use serde::Deserialize;
+use tracing::error;
+
+use self::context::Context;
+use self::session::Session;
+use self::sessions::Sessions;
+use self::sid::SessionId;
+use crate::reference::Ref;
+
+#[derive(Debug, Deserialize)]
+struct AuthRequest {
+    code: String,
+}
+
+async fn authorized<T: 'static + Send + Sync + Default>(
+    Query(AuthRequest { code, .. }): Query<AuthRequest>,
+    Extension(ctx): Extension<Ref<Context<T>>>,
+) -> Result<([(HeaderName, HeaderValue); 1], Redirect), (StatusCode, &'static str)> {
+    fn ice<E: Display>(info: &'static str) -> impl Fn(E) -> (StatusCode, &'static str) {
+        move |e: E| {
+            error!("{}: {}", info, e);
+            (StatusCode::INTERNAL_SERVER_ERROR, info)
+        }
+    }
+
+    // Get the OIDC claims.
+    let claims = ctx
+        .clone()
+        .fetch_claims(code)
+        .await
+        .map_err(ice("error fetching claims"))?;
+
+    // Get the GitHub user identifier.
+    let uid: usize = match claims.subject().split_once('|') {
+        Some(("github", uid)) => uid.parse().map_err(ice("invalid uid"))?,
+        _ => return Err(ice("invalid user type")("unknown user type")),
```

Comment:
What exactly is happening here, shouldn't it be just one of these?

Reply:
Basically, `auth0` returns `github|#`. On line 62, there is a failure case where the `uid` isn't a number. On line 63, there is a failure case where `auth0` gives us something that isn't github (which could be a misconfiguration).